talk/owt/sdk/base/win/video_capture_mf.cc

```diff
@@ -0,0 +1,478 @@
+                                .set_rotation(kVideoRotation_0)
+                                .build();
+
+  // TODO(jianlin): The timestamp returned by callback is not the same clock source
+  // of webrtc. Should we set it?
+  //captured_frame.set_ntp_time_ms(capture_time);
+
+  broadcaster_.OnFrame(captured_frame);
+
+  return;
+}
+
+HRESULT VideoCaptureMF::SetCameraOutput(
+    const VideoCaptureCapability& requested_capability) {
+  if (requested_capability_ == requested_capability) {
+    return S_OK;
+  }
+
+  // Get the best matching capability
+  VideoCaptureCapability capability;
+  int32_t capabilityIndex;
+  std::string format;
+
+  // Store the new requested size
+  requested_capability_ = requested_capability;
+  // Match the requested capability with the supported.
+  if ((capabilityIndex = info_.GetBestMatchedCapability(
+           device_unique_id_.get(), requested_capability_, capability)) < 0) {
+    return S_FALSE;
+  }
+
+  // Reduce the frame rate if possible.
+  if (capability.maxFPS > requested_capability.maxFPS) {
+    capability.maxFPS = requested_capability.maxFPS;
+  } else if (capability.maxFPS <= 0) {
+    capability.maxFPS = 30;
+  }
+
+  // Set format to capture device
+  CComPtr<IMFMediaType> media_type = nullptr;
+  HRESULT hr = reader_->GetNativeMediaType(MF_SOURCE_READER_FIRST_VIDEO_STREAM,
+      capabilityIndex, &media_type);
+  if (FAILED(hr)) {
+    RTC_LOG(LS_ERROR) << "GetNativeMediaType failed:"
+                      << std::system_category().message(hr);
+    return hr;
+  }
+
+  hr = reader_->SetCurrentMediaType(MF_SOURCE_READER_FIRST_VIDEO_STREAM, nullptr,
+                                    media_type);
+  if (FAILED(hr)) {
+    RTC_LOG(LS_ERROR) << "SetCurrentMediaType failed:"
+                      << std::system_category().message(hr);
+    return hr;
+  }
+
+  // Store the capability that is set to capture device
+  output_capability_ = capability;
+  RTC_LOG(LS_INFO) << "BestCapability Selected: width:"
+                   << capability.width << ", height:" << capability.height
+                   << ", FPS:" << capability.maxFPS
+                   << ", videoType:" << capability.videoType;
+
+  UINT32 width, height;
+  MFGetAttributeSize(media_type, MF_MT_FRAME_SIZE, &width, &height);
+  UINT32 numerator, denominator;
+  MFGetAttributeRatio(media_type, MF_MT_FRAME_RATE, &numerator, &denominator);
+  GUID subtype;
+  // Get the Video SubType for this media Type
+  media_type->GetGUID(MF_MT_SUBTYPE, &subtype);
+  if (subtype == MFVideoFormat_I420) {
+    format = "I420";
+  } else if (subtype == MFVideoFormat_IYUV) {
+    format = "IYUV";
+  } else if (subtype == MFVideoFormat_RGB24) {
+    format = "RGB24";
+  } else if (subtype == MFVideoFormat_YUY2) {
+    format = "YUY2";
+  } else if (subtype == MFVideoFormat_NV12) {
+    format = "NV12";
+  } else if (subtype == MFVideoFormat_RGB565) {
+    format = "RGB565";
+  } else if (subtype == MFVideoFormat_MJPG) {
+    format = "MJPG";
+  }
```

Comment:
What if `subtype` is not one of the type listed above?

Reply:
This would not happen. The DeviceInfoMfF::CreateCapabilityMap() will only return this set of formats. Anyway, will add some handling here in case static check tool complains.